From a1b011e919b4e8fc44a7c7ebc52330278df0e52e Mon Sep 17 00:00:00 2001
From: Dien Pham <dien.pham.ry@rvc.renesas.com>
Date: Fri, 27 May 2016 16:45:49 +0700
Subject: [PATCH 027/909] soc: renesas: rcar-sysc: Add wait 1us before turning
 off power domain

commit c20faf8c0d512d4c61e2eab3f19da3ffffa51fe5 from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

[dien.pham.ry: Apply wait 1us for only turning off power domain]

As hardware updated info,
some power domains require 1us before turning off,
e.g. A3IR, A3VP, A3VC on R-Car H3; or A3IR, A3VC on M3.

Because the waiting time is small. So, it is decided to apply
commonly for all SoC of R-Car Gen3 platform.

Signed-off-by: Dien Pham <dien.pham.ry@rvc.renesas.com>
Signed-off-by: Takeshi Kihara <takeshi.kihara.df@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/soc/renesas/rcar-sysc.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/soc/renesas/rcar-sysc.c b/drivers/soc/renesas/rcar-sysc.c
index d78d2ca72fe6..f7d12204f891 100644
--- a/drivers/soc/renesas/rcar-sysc.c
+++ b/drivers/soc/renesas/rcar-sysc.c
@@ -87,6 +87,12 @@ static int rcar_sysc_pwr_on_off(const struct rcar_sysc_ch *sysc_ch, bool on)
 	if (k == SYSCSR_RETRIES)
 		return -EAGAIN;
 
+	/* Start W/A for A3VP, A3VC, and A3IR domains */
+	if (!on && (!strcmp("a3vp", to_pd_name(sysc_ch))
+		|| !strcmp("a3ir", to_pd_name(sysc_ch))
+		|| !strcmp("a3vc", to_pd_name(sysc_ch))))
+		udelay(1);
+
 	/* Submit power shutoff or power resume request */
 	iowrite32(BIT(sysc_ch->chan_bit),
 		  rcar_sysc_base + sysc_ch->chan_offs + reg_offs);
-- 
2.17.1

